### PR TITLE
[FLINK-22911][python] Align FLIP-136 (Improve interoperability between DataStream and Table API) in PyFlink Table API

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will align FLIP-136 (Improve interoperability between DataStream and Table API) in PyFlink Table API*

## Brief change log

  - *Rework `from_data_stream`/`create_temporary_view`/`to_data_stream`*
  - *Add `to_changelog_stream`/`from_changelog_stream`*
  - *Optimize the output format of RowTypeInfo and ExternalTypeInfo*
  - *Optimize from_collection/from_elements to support RowKind and Instant*

## Verifying this change

This change added tests and can be verified as follows:

  - *tests in `test_table_environment_api.py`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/table/connector.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+from pyflink.java_gateway import get_gateway
+
+__all__ = ['ChangelogMode']
+
+
```
Comment (dianfu) [inline flink-python/pyflink/table/connector.py:22]: What about changing the module name to "changelog_mode" or "changelog"?

File: flink-python/pyflink/common/typeinfo.py [hunk 2]
```diff
@@ -182,6 +185,28 @@ def BIG_INT_TYPE_INFO():
     def BIG_DEC_TYPE_INFO():
         return BasicTypeInfo(BasicType.BIG_DEC)
 
+    @staticmethod
+    def INSTANT_TYPE_INFO():
+        return InstantTypeInfo(BasicType.INSTANT)
+
+
+class InstantTypeInfo(BasicTypeInfo):
```
Comment (dianfu) [inline flink-python/pyflink/common/typeinfo.py:193]: Add documentation about instant type in https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/dev/python/datastream/data_types/#supported-data-types

File: flink-python/pyflink/common/types.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
 
 from typing import List
 
-__all__ = ['Row', 'RowKind']
+__all__ = ['Row', 'RowKind', 'Instant']
```
Comment (dianfu) [inline flink-python/pyflink/common/types.py:23]: What about move Instant to pyflink/common/time.py

File: flink-python/pyflink/fn_execution/coder_impl_fast.pyx [hunk 4]
```diff
@@ -262,9 +263,10 @@ cdef class IterableCoderImpl(LengthPrefixBaseCoderImpl):
         self._separated_with_end_message = separated_with_end_message
 
     cpdef encode_to_stream(self, value, LengthPrefixOutputStream output_stream):
-        for item in value:
-            self._field_coder.encode_to_stream(item, self._data_out_stream)
-            self._write_data_to_output_stream(output_stream)
+        if value:
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/coder_impl_fast.pyx:266]: What's the purpose of this change?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/utils/python/PythonTableUtils.scala [hunk 5]
```diff
@@ -147,8 +147,8 @@ object PythonTableUtils {
 
     case _ if dataType == org.apache.flink.api.common.typeinfo.Types.INSTANT =>
       (obj: Any) => nullSafeConvert(obj) {
-        case c: Long => Instant.ofEpochMilli(c / 1000)
-        case c: Int => Instant.ofEpochMilli(c.toLong / 1000)
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/utils/python/PythonTableUtils.scala:151]: What's the purpose of this change?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/utils/python/PythonTableUtils.scala [hunk 6]
```diff
@@ -216,18 +216,19 @@ object PythonTableUtils {
       (obj: Any) => nullSafeConvert(obj) {
         case c if c.getClass.isArray =>
           val r = c.asInstanceOf[Array[_]]
-          if (r.length != rowType.getFieldTypes.length) {
+          if (r.length - 1 != rowType.getFieldTypes.length) {
```
Comment (dianfu) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/utils/python/PythonTableUtils.scala:219]: What's the purpose of this change? Is this fixing a bug?

**Review by dianfu — COMMENTED** [review]
@HuangXingBo Thanks a lot for the PR.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/utils/python/PythonTableUtils.scala (hunk 6 above)
Comment (HuangXingBo) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/utils/python/PythonTableUtils.scala:219]: This is adding a `from_collection` support for RowKind, which is a new feature

File: flink-python/pyflink/fn_execution/coder_impl_fast.pyx (hunk 4 above)
Comment (HuangXingBo) [reply flink-python/pyflink/fn_execution/coder_impl_fast.pyx:266]: It is a hotfix for IterableCoderImpl when input value is a None after we reconstructiing coders

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I have addressed the comments at the corresponding commits.

File: flink-python/pyflink/common/time.py [hunk 7]
```diff
@@ -17,6 +17,8 @@
 ################################################################################
 from pyflink.java_gateway import get_gateway
 
+__all__ = ['Instant']
```
Comment (dianfu) [inline flink-python/pyflink/common/time.py:20]: Nit:
- Keep it consistent for both Duration and Instant
- What about also declare Instant in __init__.py?

**Review by dianfu — APPROVED** [review]
LGTM. Just left one minor comment.
